core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeUnionRule.java

```diff
@@ -90,9 +96,42 @@ public EnumerableMergeUnionRule(Config config) {
       }
     }
 
+    final RelBuilder builder = call.builder();
     final List<RelNode> inputs = new ArrayList<>(unionInputsSize);
     for (RelNode input : union.getInputs()) {
-      final RelNode newInput = sort.copy(sort.getTraitSet(), input, collation, null, inputFetch);
+      // Check if type collations match, otherwise store it in this map to generate a cast later
+      // to guarantee that all inputs will be sorted in the same way
+      final Map<Integer, RelDataType> castMap = new HashMap<>();
```

Comment:
How about using a bit set instead of a map? The value is only used once and can easily be looked up again.

Reply:
Makes sense. Change applied.

Reply:
The union.getRowType().getFieldList() could be factored out of the for-loop. I guess that it does not impact the performance much, so up to you to decide.

Reply:
ok, done.